docs/YOSEMITE.md

```diff
@@ -0,0 +1,103 @@
+# Yosemite
+> This module encapsulates the application business logic, and interacts with the Networking and Storage layers.   
+
+## High-level class diagram
+![Yosemite high level class diagram](images/yosemite_general.png)
+
+## Application state. StoresManagerState
+Business logic is broken down into different subclasses of   `Store`   A `Store` encapsulates the business logic related to one and only one of the domain level concerns (i.e. Notifications, Orders, Shipment…)
+
+Some of these concerns require a valid user session (a user to be logged into the system) while some others don’t require a session.  To handle that, `Store`s are aggregated by a `StoresManager`.
+
+`StoresManager` is a state machine, that manages two states. Those states are implementations of the `StoresManagerState` protocol. Currently, we have two implementations, `AuthenticateState` and `DeauthenticateState`
+
+Each of the implementations of the `StoresManagerState` aggregates a collection of subclasses of `Store` protocol. That’s how we enforce certain concerns (i.e. Orders) to require users to be authenticated.
+
+`StoresManager` takes care of the transition from the authenticated (`AuthenticateState`) to the deauthenticated (`Deauthenticate`) state, while executing the necessary side effects of those transitions (for example, clearing the user data locally stored)
+
+`StoresManager` is a singleton. 
+
+The `AuthenticateState`aggregates a reference to `Dispatcher`. This is an important piece of the system, as it is responsible for executing the actions declared in the different Stores. Therefore we will discuss it in depth later on.
+
+## Domain level concerns. Store
+As mentioned in the previous section, the business logic pertaining each domain level concern is encapsulated in a separate subclass of `Store` .
+
+`Store` is an implementation of the `ActionsProcessor` protocol. We will discuss this protocol later, but for now, let’s say it is _something that can process an action_.
+
+`Store` is injected with a reference to a `Dispatcher`, an implementation of the  `StorageManagerType`  protocol (the storage stack) and an implementation of the `Network` protocol (the networking stack).
+
+As a quick aside, please note this pattern: dependencies are usually injected.
+
+`Store` is a lightweight implementation of the Template Method Pattern. It enforces its subclasses to implement a method  (called `registerSupportedActions`) that will register a collection of _supported actions_ with the `Dispatcher` and to implement a method that deals with incoming actions.  We will discuss what actions are and how they are modelled a bit later.
```

Comment:
Can you add a link to your favorite description of the Template Method Pattern? I don't know that design pattern.

Reply:
Ouch, I totally forgot to add it, sorry!